### PR TITLE
CSHARP-2162: Add disableMD5 option and deprecate MD5 field

### PR DESCRIPTION
**Pull request by vincentkam**

### CODE REVIEW

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs [hunk 1]
```diff
@@ -25,6 +25,7 @@ public class GridFSBucketOptions
     {
         // fields
         private string _bucketName;
+        private bool _disableMD5 = false;
```
Comment (rstam) [inline src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: Keep in alphabetical order?

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs [hunk 2]
```diff
@@ -99,6 +102,18 @@ public int ChunkSizeBytes
                 _chunkSizeBytes = value;
             }
         }
+        
+        /// <summary>
+        /// Gets or set whether or not GridFS will compute MD5 checksums of uploaded files.
```
Comment (rstam) [inline src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: While true in a general sense this sounds backwards. How about:

Gets or sets whether to disable MD5 checksum computation when uploading a GridFS file.

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs [hunk 3]
```diff
@@ -184,6 +200,7 @@ public ImmutableGridFSBucketOptions(GridFSBucketOptions other)
         {
             Ensure.IsNotNull(other, nameof(other));
             _bucketName = other.BucketName;
+            _disableMD5 = other.DisableMD5;
```
Comment (rstam) [inline src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: Keep in alphabetical order?

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs [hunk 4]
```diff
@@ -212,6 +229,18 @@ public int ChunkSizeBytes
         {
             get { return _chunkSizeBytes; }
         }
+        
+        /// <summary>
+        /// Gets or set whether or not GridFS will compute MD5 checksums of uploaded files.
```
Comment (rstam) [inline src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: If you change the other description change this one too.

File: tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs [hunk 5]
```diff
@@ -139,6 +139,16 @@ public void constructor_with_no_arguments_should_initialize_instance_with_defaul
             result.WriteConcern.Should().BeNull();
         }
 
+        [Fact]
+        public void DisableMD5_get_set_should_return_expected_result()
```
Comment (rstam) [inline tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs]: Might be worth making this a theory and testing with both true and false.

File: tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs [hunk 6]
```diff
@@ -264,11 +284,22 @@ public void Defaults_get_should_return_expected_result()
...
             result.BucketName.Should().Be("fs");
             result.ChunkSizeBytes.Should().Be(255 * 1024);
+            result.DisableMD5.Should().Be(false);
             result.ReadConcern.Should().BeNull();
             result.ReadPreference.Should().BeNull();
             result.WriteConcern.Should().BeNull();
         }
 
+        [Fact]
+        public void DisableMD5_get_and_set_should_return_expected_result()
```
Comment (rstam) [inline tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs]: Might be worth making this a theory and testing with both true and false.

File: src/MongoDB.Driver.GridFS/GridFSForwardOnlyUploadStream.cs [hunk 7]
```diff
@@ -76,7 +78,8 @@ internal class GridFSForwardOnlyUploadStream<TFileId> : GridFSUploadStream<TFile
             _batchSize = batchSize;
 
             _batch = new List<byte[]>();
-            _md5 = IncrementalMD5.Create();
+            _md5 = disableMD5 ? null : IncrementalMD5.Create();
```
Comment (rstam) [inline src/MongoDB.Driver.GridFS/GridFSForwardOnlyUploadStream.cs:81]: CreateWriteBatchChunkDocuments in this same file accesses _md5 without checking if it is null.

Not sure why that didn't fail any tests with a NullReferenceException.

p.s. I think it's because you tested disableMD5 with a zero length file.

File: src/MongoDB.Driver.GridFS/GridFSUploadOptions.cs [hunk 8]
```diff
@@ -85,6 +87,18 @@ public string ContentType
             get { return _contentType; }
             set { _contentType = Ensure.IsNullOrNotEmpty(value, nameof(value)); }
         }
+        
+        /// <summary>
+        /// Gets or sets whether or not MD5 is disabled.
```
Comment (rstam) [inline src/MongoDB.Driver.GridFS/GridFSUploadOptions.cs]: Keep this description consistent with other descriptions of DisableMD5.

**Review by rstam — CHANGES_REQUESTED** [review]
Looks good overall with a few minor comments.

File: src/MongoDB.Driver.GridFS/GridFSForwardOnlyUploadStream.cs (hunk 7 above)
Comment (vincentkam) [reply src/MongoDB.Driver.GridFS/GridFSForwardOnlyUploadStream.cs:81]: Thank you, fixed!

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: Fixed!

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs (hunk 2 above)
Comment (vincentkam) [reply src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: Ah, true. I'll use your summary.

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs (hunk 3 above)
Comment (vincentkam) [reply src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: Fixed.

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs (hunk 4 above)
Comment (vincentkam) [reply src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: Updated to use new description.

File: src/MongoDB.Driver.GridFS/GridFSUploadOptions.cs (hunk 8 above)
Comment (vincentkam) [reply src/MongoDB.Driver.GridFS/GridFSUploadOptions.cs]: Fixed.

File: tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs (hunk 5 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs]: Done. Also tweaked the name to match precedent (get_and_set).

File: tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs (hunk 6 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.GridFS.Tests/GridFSBucketOptionsTests.cs]: Done.

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs [hunk 9]
```diff
@@ -99,6 +102,18 @@ public int ChunkSizeBytes
                 _chunkSizeBytes = value;
             }
         }
+        
+        /// <summary>
+        /// Gets or sets whether to disable MD5 checksum computation when uploading a GridFS file.
+        /// </summary>
+        /// <value>
+        /// Whether or not MD5 checksum computation is disabled when uploading a GridFS file.
```
Comment (rstam) [inline src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: This text is out of sync with the summary (but only *very* slightly).

Doesn't need to be updated unless you want to.

**Review by rstam — APPROVED** [review]

**Comment by rstam** [comment]
Note: don't forget to check whether a rebase is needed.

File: tests/MongoDB.Driver.GridFS.Tests/Specifications/gridfs/tests/upload.yml [hunk 10]
```diff
@@ -156,3 +156,18 @@ tests:
                     { insert : "expected.chunks", documents : [
                         { _id : "*actual", files_id : "*result", n : 0, data : { $hex : "11" } }
                     ] }
+    -
+        description: "Upload when length is 0 sans MD5"
```
Comment (rstam) [inline tests/MongoDB.Driver.GridFS.Tests/Specifications/gridfs/tests/upload.yml:160]: I know I already lgtm'ed this pull request, but I realized after the fact that we should probably add one more test just like this one but with a non-zero length source.

Comment (rstam) [reply tests/MongoDB.Driver.GridFS.Tests/Specifications/gridfs/tests/upload.yml:160]: @jyemin can you comment on this?

Perhaps we just need to get this committed ASAP.

Comment (jyemin) [reply tests/MongoDB.Driver.GridFS.Tests/Specifications/gridfs/tests/upload.yml:160]: Will it exercise additional code paths?

Comment (rstam) [reply tests/MongoDB.Driver.GridFS.Tests/Specifications/gridfs/tests/upload.yml:160]: Yes. It would have revealed the NullReferenceException bug that I caught visually in an earlier commit.

Comment (jyemin) [reply tests/MongoDB.Driver.GridFS.Tests/Specifications/gridfs/tests/upload.yml:160]: I think a test should be added.  Shouldn't take long.

Comment (vincentkam) [reply tests/MongoDB.Driver.GridFS.Tests/Specifications/gridfs/tests/upload.yml:160]: Added!

File: src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs (hunk 9 above)
Comment (vincentkam) [reply src/MongoDB.Driver.GridFS/GridFSBucketOptions.cs]: Tweaking as discussed!

**Review by rstam — APPROVED** [review]
